### PR TITLE
[Vectorized][UDF] support java-udaf

### LINKED ISSUE #8389: [Feature] Support Java UDF
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

## Backgroup
Doris currently supports both [Native UDF](http://doris.incubator.apache.org/zh-CN/extending-doris/udf/native-user-defined-function.html) and [Remote UDF](http://doris.incubator.apache.org/zh-CN/extending-doris/udf/remote-user-defined-function.html) for user-defined functions.

Native UDF is written in C++ and has the best performance, but it is more difficult to write and debug, and may be limited by some system library versions(such as libc) that may not be compatible after upgrade.

Remote UDF solves the language problem very well, and in theory, UDF logic can be written in any language. But the disadvantage is that users need to implement their own high-performance UDF Service, and the efficiency is not good because of the RPC problem.

## Motivation

Support for Java UDF is motivated by the following points:

1. the Java language is widely used, rich support for related library functions, and a low threshold for development and debugging.
2. Some big data ecologies such as Hive, Spark, etc. already exist a large number of ready-made UDF, we need to be compatible with these udf to reduce the migration costs of users.
3. If there is bug in UDF, it will not cause BE crash

## How to implement

Need to start the JVM and call the relevant UDF logic from the BE side by way of JNI.

## Referance:

Remote UDF: #7519
Lua UDF:#5979

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

Issue Number: close #8389

## Problem Summary:
This feature is used to support Java udaf. Users can create aggregate functions by self,
For example, we create Java udaf function by code below.

```
CREATE AGGREGATE FUNCTION simple_add (int) RETURNS int PROPERTIES (
    "file"="file:///pathTo/java-udaf.jar",
    "symbol"="org.apache.doris.udf.SimpleAdd",
    "type"="JAVA_UDF"
);
```

**file:** indicate where user file is.
**symbol:** for java udaf means udaf class in this jar.
**type:** indicate this function is a java udf.

Users may have some necessary conditions 
when implementing aggregate functions by self,
for example:

```
class SimpleAdd {
    public static class State { 
        some variables: 
        example count agg: maybe a long type counter;
    }

    public State create() {
        ....... if need do some.......
        return new State();
    }

    public void destroy(State s) {
        ....... if need do some destroy work.......
    }

    public void add(State s, Type val, [Type val2, .....]) {
        do update work here;
        like count agg function: you need plus 1
   }

    public void serialize(State s, DataOutputStream out) {
        put you add result into buffer to serialize
   }

  public void deserialize(State state, DataInputStream in) {
      deserialize get data from buffer before you put
  }

    public void merge(State s, State other) {
       merge result from other state
    }

    public Type getValue(State s) {
          return finally result
   }
}

```

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+            }
+            RETURN_ERROR_IF_EXC(env);
+            RETURN_IF_ERROR(JniUtil::LocalToGlobalRef(env, executor_obj, &executor_obj));
+            first_init = false;
+        }
+        return Status::OK();
+    }
+
+    void add(const IColumn** columns, size_t row_num, const DataTypes& argument_types) const {
+        JNIEnv* env = nullptr;
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:131]: we can wrap a function `JNIEnv* get_env(string when)`

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        if (rhs.first_init) {
+            return;
+        }
+        JNIEnv* env = nullptr;
+        Status ret_code = JniUtil::GetJNIEnv(&env);
+        if (env == nullptr || !ret_code.ok()) {
+            LOG(WARNING) << "Java-Udaf get error when merge: " << ret_code.get_error_msg();
+        }
+        serialize_data = rhs.serialize_data;
+        long len = serialize_data.length();
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:175]: Should `long` be used here? `long` has different meaning in `c++` and `java`.

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+    String get_name() const override { return fn_.name.function_name; }
+
+    DataTypePtr get_return_type() const override { return return_type_; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena*) const override {
+        Status status = data(place).init_udaf(fn_);
+        if (!status.ok()) {
+            LOG(WARNING) << " in add pressore  of init udaf has error: " << status.get_error_msg();
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:352]: multiple consecutive spaces

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    } else if (data_col.is_numeric() || data_col.is_column_decimal()) {                      \
+        *output_value_buffer = reinterpret_cast<int64_t>(data_col.get_raw_data().data);      \
+        env->CallNonvirtualBooleanMethod(executor_obj, executor_cl_, executor_result_id_,    \
+                                         to.size() - 1);                                     \
+    } else {                                                                                 \
+        LOG(WARNING) << "Java UDAF doesn't support return type: " << result_type->get_name() \
+                     << "$0 now !";                                                          \
+    }
+#endif
+            EVALUATE_JAVA_UDAF;
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:254]: Do you just want a lambda function?

**Comment by morningman** [comment]
Hi @zhangstar333 , could you explain more about what does this PR do?
Because we already supported part of JAVA udf feature, so you need to explain what's new in your PR.

**Comment by HappenLee** [comment]
need also modify the doc，please

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(fn_));
+        this->data(place).read(buf);
+    }
+
+    void insert_result_into(ConstAggregateDataPtr __restrict place, IColumn& to) const override {
+        this->data(place).get(to, return_type_);
+    }
+
+private:
+    TFunction fn_;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:356]: `_fn`

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    }
+
+    String get_name() const override { return fn_.name.function_name; }
+
+    DataTypePtr get_return_type() const override { return return_type_; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena*) const override {
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(fn_));
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:320]: why not do init in `create` function and delete the `first_init`?

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        return Status::OK();
+    }
+
+    void read(BufferReadable& buf) {
+        read_binary(first_init, buf);
+        read_binary(serialize_data, buf);
+    }
+
+    Status get(IColumn& to, const DataTypePtr& result_type) const {
+        to.insert_default();
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:186]: rethink the `insert_default()` of column string

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+                                            argument_types[arg_idx]->get_name()));
+            }
+        }
+        env->CallNonvirtualVoidMethod(executor_obj, executor_cl, executor_add_id, row_num_start,
+                                      row_num_end);
+        return Status::OK();
+    }
+
+    Status merge(const AggregateJavaUdafData& rhs) {
+        if (rhs.first_init) {
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:156]: here is werid of `first_init`, rethink the logic

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+                UdfUtils.copyMemory(bytes, UdfUtils.BYTE_ARRAY_OFFSET, null,
+                        UdfUtils.UNSAFE.getLong(null, outputBufferPtr) + offset - bytes.length - 1, bytes.length);
+                return true;
+            default:
+                throw new UdfRuntimeException("Unsupported return type: " + retType);
+        }
+    }
+
+    private void allocateInputObjects(long row) throws UdfRuntimeException {
+        inputObjects = new Object[argTypes.length];
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:343]: why not return inputObjects?

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+                    if (UdfUtils.UNSAFE.getLong(null, UdfUtils.getAddressAtOffset(inputNullsPtrs, i)) == -1
+                            || UdfUtils.UNSAFE.getByte(null,
+                                    UdfUtils.UNSAFE.getLong(null, UdfUtils.getAddressAtOffset(inputNullsPtrs, i)) + row)
+                            == 0) {
+                        inputArgs[i + 1] = inputObjects[i];
+                    } else {
+                        inputArgs[i + 1] = null;
+                    }
+                }
+                inputArgs[0] = stateObj;
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:152]: only need set one time?

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+                    UdfUtils.copyMemory(null, base, bytes, UdfUtils.BYTE_ARRAY_OFFSET, numBytes);
+                    inputObjects[i] = new String(bytes, StandardCharsets.UTF_8);
+                    break;
+                default:
+                    throw new UdfRuntimeException("Unsupported argument type: " + argTypes[i]);
+            }
+        }
+    }
+
+    private URLClassLoader getClassLoader(String jarPath) throws MalformedURLException {
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:429]: same as method of `UdfExecutor` only keep one, please

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+            env->CallNonvirtualBooleanMethod(executor_obj, executor_cl, executor_result_id,
+                                             to.size() - 1);
+        }
+        return Status::OK();
+    }
+
+private:
+    static const int32_t INITIAL_RESERVED_BUFFER_SIZE = 1024;
+    // TODO: we need a heuristic strategy to increase buffer size for variable-size output.
+    static inline int32_t IncreaseReservedBufferSize(int n) {
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:242]: Move this function to another common file and call it in UDAF and UDF

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h (hunk 6 above)
Comment (zhangstar333) [reply be/src/vec/aggregate_functions/aggregate_function_java_udaf.h]: Have test in create function do init operator, but it's unable to get jvm

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+
+    static AggregateFunctionPtr create(const TFunction& fn, const DataTypes& argument_types,
+                                       const Array& parameters, const DataTypePtr& return_type) {
+        return std::make_shared<AggregateJavaUdaf>(fn, argument_types, parameters, return_type);
+    }
+
+    void create(AggregateDataPtr __restrict place) const override {
+        new (place) Data(argument_types.size());
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:307]: use `RETURN_IF_ERROR`

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    String get_name() const override { return _fn.name.function_name; }
+
+    DataTypePtr get_return_type() const override { return _return_type; }
+
+    // TODO: here calling add operator maybe only hava done one row, this performance may be poorly
+    // so it's possible to maintain a hashtable in FE, the key is place address, value is the object
+    // then we can calling add_bacth function and calculate the whole batch at once,
+    // and avoid calling jni multiple times.
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:317]: Padding a space at the beginning for each commented line below `TODO`

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+             Arena*) const override {
+        this->data(place).add(columns, row_num, row_num + 1, argument_types);
+    }
+
+    // TODO: Here we calling method by jni, And if we get a thrown from FE,
+    // But can't let user known the error, only return directly and output error to log file.
+    void add_batch_single_place(size_t batch_size, AggregateDataPtr place, const IColumn** columns,
+                                Arena* arena) const override {
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:328]: use `RETURN_IF_ERROR`

Comment (Gabriel39) [reply be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:328]: Is it necessary to `init_udaf` for each time to call `add_batch_single_place`? Seems you have called it in `create`.

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+
+    std::unique_ptr<int64_t[]> input_values_buffer_ptr;
+    std::unique_ptr<int64_t[]> input_nulls_buffer_ptr;
+    std::unique_ptr<int64_t[]> input_offsets_ptrs;
+    std::unique_ptr<int64_t> output_value_buffer;
+    std::unique_ptr<int64_t> output_null_value;
+    std::unique_ptr<int64_t> output_offsets_ptr;
+    std::unique_ptr<int64_t> output_intermediate_state_ptr;
+
+    bool first_init;
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:284]: I think this variable name is a bit curious. `init` also means `first`. Maybe rename to `inited` or `first_called` ?

File: be/src/vec/aggregate_functions/aggregate_function_java_udaf.h [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+    Status write(BufferWritable& buf) {
+        write_binary(first_init, buf);
+        JNIEnv* env = nullptr;
+        RETURN_NOT_OK_STATUS_WITH_WARN(JniUtil::GetJNIEnv(&env), "Java-Udaf write function");
+        // TODO: Here get a byte[] from FE serialize, and then allocate the same length bytes to
+        // save it in BE, Because i'm not sure there is a way to use the byte[] not allocate again.
+        jbyteArray arr = (jbyteArray)(env->CallNonvirtualObjectMethod(executor_obj, executor_cl,
+                                                                      executor_serialize_id));
+        int len = env->GetArrayLength(arr);
+        serialize_data.resize(len);
```
Comment (Gabriel39) [inline be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:177]: call `serialize_data` before using it

**Review by Gabriel39 — CHANGES_REQUESTED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**